DynamoDBPaxos.cpp

```
@@ -0,0 +1,439 @@
+    config.scheme = Aws::Http::Scheme::HTTP;
+    config.endpointOverride = "localhost:8000";
+
+    // Set private fields
+    this->my_dc = dc;
+    this->node_id = node_id;
+    this->dcs = dcs;
+    this->table = Aws::MakeShared<Aws::DynamoDB::DynamoDBClient>(nullptr, credential, config);
+
+    // check if table exists
+    DescribeTableRequest req;
+    req.SetTableName(dc);
+    DescribeTableOutcome out = this->table->DescribeTable(req);
+    if (!out.IsSuccess()) {
+        // create the table
+        Aws::Vector<AttributeDefinition> attribute_definitions = {getAttributeDefinition("UserId", "N"),
+                                                                  getAttributeDefinition("BlobId", "N")};
+        Aws::Vector<KeySchemaElement> key_schemas = {getKeySchema("UserId", "HASH"),
+                                                     getKeySchema("BlobId", "RANGE")};
+        CreateTableRequest createTableRequest;
+        createTableRequest.WithTableName(dc)
+                .WithAttributeDefinitions(attribute_definitions)
+                .WithKeySchema(key_schemas)
+                .WithProvisionedThroughput(getProvisionedThroughput(1, 1));
+        CreateTableOutcome outcome = this->table->CreateTable(createTableRequest);
+        if (outcome.IsSuccess()) {
+            printf("Giza Node %d Successfully Created Table %s \n", node_id, dc.c_str());
+        } else {
+            printf("Giza Node %d Failed Created Table %s with error  %s \n", node_id, dc.c_str()
+                    ,outcome.GetError().GetMessage().c_str());
+        }
+    }
+};
+
+int DBPaxosKV::Put(int user_id, int blob_id, string md) {
+    int version = getVersionNumber(user_id, blob_id);
+    printf("DC: %s Node: %d Put(%d-%d, %s) Version: %d\n", my_dc.c_str(), node_id, user_id, blob_id, md.c_str(), version + 1);
+    string result;
+    do {
+        version ++;
+        result = runPaxos(user_id, blob_id, md, version);
+        updateVersionNumber(user_id, blob_id, version);
```

This is hard to read. You did not check the result after running Paxos. 
You don't want to updateVersion number if Paxos is not successful. I know updateVersionNumber will silently drop the update if Paxos is not successful. but it's better to make your code logic explicit.